### PR TITLE
Allow narrowing R in more Sink combinators

### LINKED ISSUE #603: Allow narrowing of `R` in Sink composition

### PR DESCRIPTION
**Pull request by iravid**
Resolves #603

### CODE REVIEW

File: core/shared/src/main/scala/scalaz/zio/stream/Sink.scala [hunk 1]
```diff
@@ -428,13 +441,13 @@ object Sink {
 
   type Step[+S, +A0] = Step.Step[S, A0]
 
-  final def more[R, E, A0, A, B](end: IO[E, B])(input: A => Sink[R, E, A0, A, B]): Sink[R, E, A0, A, B] =
-    new Sink[R, E, A0, A, B] {
-      type State = Option[(Sink[R, E, A0, A, B], Any)]
+  final def more[R, R1 <: R, E, A0, A, B](end: ZIO[R1, E, B])(input: A => Sink[R, E, A0, A, B]): Sink[R1, E, A0, A, B] =
```
Comment (wi101) [inline core/shared/src/main/scala/scalaz/zio/stream/Sink.scala]: `input: A => Sink[R1, E, A0, A, B]` ?

File: core/shared/src/main/scala/scalaz/zio/stream/Sink.scala [hunk 2]
```diff
@@ -621,14 +634,14 @@ object Sink {
...
 
-        val initial: ZIO[R, E1, Step[State, Nothing]] =
+        val initial: ZIO[R1, E1, Step[State, Nothing]] =
           l.zipWithPar(r) { (l, r) =>
             Step.both(Step.leftMap(l)(eitherToSide), Step.leftMap(r)(eitherToSide))
           }
 
-        def step(state: State, a: A): ZIO[R, E1, Step[State, A]] = {
+        def step(state: State, a: A): ZIO[R1, E1, Step[State, A]] = {
           val leftStep: ZIO[R, Nothing, Step[Side[E, self.State, B], A]] =
```
Comment (wi101) [inline core/shared/src/main/scala/scalaz/zio/stream/Sink.scala]: ` val leftStep: ZIO[R1, ..` ?

File: core/shared/src/main/scala/scalaz/zio/stream/Sink.scala [hunk 3]
```diff
@@ -139,15 +139,15 @@ trait Sink[-R, +E, +A0, -A, +B] { self =>
   final def filterNotM[E1 >: E, A1 <: A](f: A1 => IO[E1, Boolean]): Sink[R, E1, A0, A1, B] =
     filterM(a => f(a).map(!_))
 
-  final def contramapM[E1 >: E, C](f: C => IO[E1, A]): Sink[R, E1, A0, C, B] =
-    new Sink[R, E1, A0, C, B] {
+  final def contramapM[R1 <: R, E1 >: E, C](f: C => ZIO[R1, E1, A]): Sink[R1, E1, A0, C, B] =
+    new Sink[R1, E1, A0, C, B] {
       type State = self.State
```
Comment (wi101) [inline core/shared/src/main/scala/scalaz/zio/stream/Sink.scala]: 👍

File: core/shared/src/main/scala/scalaz/zio/stream/Sink.scala (hunk 2 above)
Comment (iravid) [reply core/shared/src/main/scala/scalaz/zio/stream/Sink.scala]: `leftStep` is actually the step from the current sink, which uses `R`. `rightStep` is from the other sink which uses a narrower `R1`

File: core/shared/src/main/scala/scalaz/zio/stream/Sink.scala (hunk 1 above)
Comment (iravid) [reply core/shared/src/main/scala/scalaz/zio/stream/Sink.scala]: So the input function yields sinks that use `R`, and the ending value `end` uses `R1`

File: core/shared/src/main/scala/scalaz/zio/stream/Sink.scala (hunk 2 above)
Comment (wi101) [reply core/shared/src/main/scala/scalaz/zio/stream/Sink.scala]: it makes sense, thanks
LGTM

**Pull request merged**